File: Leetcode/Python3/104. Maximum Depth of Binary Tree/solution.py
# https://leetcode.com/problems/maximum-depth-of-binary-tree/

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def dfs(self, root, depth):
        if root:
            return max(self.dfs(root.left, depth + 1), self.dfs(root.right, depth + 1))
        return depth
    def maxDepth(self, root: Optional[TreeNode]) -> int:    # Runtime: 89 ms
        return self.dfs(root, 0)
    
    def maxDepth(self, root: Optional[TreeNode]) -> int:    # Runtime: 56 ms
        def dfs(root, depth):
            if root:
                return max(dfs(root.left, depth + 1), dfs(root.right, depth + 1))
            return depth
        return dfs(root, 0)

    

File: Leetcode/Java/104. Maximum Depth of Binary Tree/solution.java
// https://leetcode.com/problems/maximum-depth-of-binary-tree/description/

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int maxDepth(TreeNode root) { // Tree level BFS
        if (root == null){
            return 0;
        }
        Queue<TreeNode> queue = new LinkedList<>();
        queue.add(root);
        int depth = 0;

        // Get all each tree level node
        while (!queue.isEmpty()){
            depth += 1;
            int levelSize = queue.size();

            for (int i = 0; i < levelSize; i++){
                TreeNode node = queue.poll();
                
                if (node.left != null){
                    queue.add(node.left);
                }
                if (node.right != null){
                    queue.add(node.right);
                }
            }
        }

        return depth;
    }
}